test/couch_epi_tests.erl

```diff
@@ -0,0 +1,491 @@
+    end);
+ensure_notified_when_changed(Module, #ctx{key = Key} = Ctx) ->
+    ?_test(begin
+        subscribe(Ctx, test_app, Key),
+        update(Module, Ctx),
+        timer:sleep(200),
+        ExpectedData = lists:usort([
+            {[complex, key, 1], [{type, counter}, {desc, updated_foo}]},
+            {[complex, key, 2], [{type, counter}, {desc, bar}]}
+        ]),
+        Result = get(Ctx, is_called),
+        ?assertMatch({ok, {test_app, Key, {data, _}, {data, _}}}, Result),
+        {ok, {test_app, Key, {data, OldData}, {data, Data}}} = Result,
+        ?assertMatch(ExpectedData, lists:usort(Data)),
+        ?assertMatch(
+            [{[complex, key, 1], [{type, counter}, {desc, foo}]}],
+            lists:usort(OldData))
+    end).
+
+ensure_not_notified_when_no_change(_Module, #ctx{key = Key} = Ctx) ->
+    ?_test(begin
+        subscribe(Ctx, test_app, Key),
+        timer:sleep(200),
+        ?assertMatch(error, get(Ctx, is_called))
+    end).
+
+ensure_not_notified_when_unsubscribed(Module, #ctx{key = Key} = Ctx) ->
+    ?_test(begin
+        SubscriptionId = subscribe(Ctx, test_app, Key),
+        couch_epi:unsubscribe(SubscriptionId),
+        timer:sleep(100),
+        update(Module, Ctx),
+        timer:sleep(200),
+        ?assertMatch(error, get(Ctx, is_called))
+    end).
+
+ensure_apply_is_called(Opts, #ctx{handle = Handle, kv = KV, key = Key} = Ctx) ->
+    ?_test(begin
+        couch_epi:apply(Handle, Key, inc, [KV, 2], Opts),
+        maybe_wait(Opts),
+        ?assertMatch({ok, _}, get(Ctx, inc1)),
+        ?assertMatch({ok, _}, get(Ctx, inc2)),
+        ok
+    end).
+
+check_pipe(#ctx{handle = Handle, kv = KV, key = Key}) ->
+    ?_test(begin
+        Result = couch_epi:apply(Handle, Key, inc, [KV, 2], [pipe]),
+        ?assertMatch([KV, 4], Result),
+        ok
+    end).
+
+check_broken_pipe(#ctx{handle = Handle, kv = KV, key = Key} = Ctx) ->
+    ?_test(begin
+        Result = couch_epi:apply(Handle, Key, fail, [KV, 2], [pipe, ignore_errors]),
+        ?assertMatch([KV, 3], Result),
+        ?assertMatch([3, check_error], pipe_state(Ctx)),
+        ok
+    end).
+
+ensure_fail_pipe(#ctx{handle = Handle, kv = KV, key = Key}) ->
+    ?_test(begin
+        ?assertThrow(check_error,
+            couch_epi:apply(Handle, Key, fail, [KV, 2], [pipe])),
+        ok
+    end).
+
+ensure_fail(#ctx{handle = Handle, kv = KV, key = Key}) ->
+    ?_test(begin
+        ?assertThrow(check_error,
+            couch_epi:apply(Handle, Key, fail, [KV, 2], [])),
+        ok
+    end).
+
+ensure_unsubscribe_when_caller_die(_Module, #ctx{key = Key} = Ctx) ->
+    ?_test(begin
+        spawn(fun() ->
+            subscribe(Ctx, test_app, Key)
+        end),
+        timer:sleep(200),
+        ?assertMatch(error, get(Ctx, is_called))
+    end).
+
+
+pipe_state(Ctx) ->
+    Trace = [get(Ctx, inc1), get(Ctx, inc2)],
+    lists:usort([State || {ok, State} <- Trace]).
+
+check_dump(_Module, #ctx{handle = Handle}) ->
+    ?_test(begin
```

Comment:
`?_test(?assertX(...))` is [actually](https://github.com/erlang/otp/blob/maint/lib/eunit/include/eunit.hrl#L123-L138) `?_assertX(...)`